### ISSUE TITLE
Outgoing edges of nodes should be sorted based on when a call `.stream()` created them

### ISSUE THREAD

**Issue by Depau** [kkroening/ffmpeg-python#64]
```python
i = ffmpeg.input("f1.mp4")
ref = ffmpeg.input("f2.mp4")
# BTW `filter_multi_output` is not in __all__, needs #65
s2ref = ffmpeg.filter_([i, ref], "scale2ref").node
scaled = s2ref[0]
ref2 = s2ref[1]
```
https://ffmpeg.org/ffmpeg-all.html#scale2ref

[![Screenshot_from_2018-01-26_14-33-23.png](https://s10.postimg.org/k87rpqofd/Screenshot_from_2018-01-26_14-33-23.png)](https://postimg.org/image/f9k9b7kmd/)

You would expect the output of `scale2ref` labeled 0 (`scaled`) to be the first output of `scale2ref` (because `.stream()` was called first), and 1 (`ref2`) to be the second. However, this doesn't always happen: it's totally up to `topo_sort()`.

I'll try to fix it and send a PR.

**Comment by Depaulicious** [comment]
There are a number of ways sorting can be ensured; the ones that came to my mind involve storing some time-related value in `Stream` instances, in `__init__`, like:

- `time.time()` (wall clock time since Epoch)
  - pros: simple, no changes in the API
  - cons: issues will occur if system time is adjusted towards the past during computation; depends on `time`
- `time.clock()` (process CPU time)
  - pros: simple, no changes in the API
  - cons: depends on `time`; issues with `multiprocessing`
- Instance counter in Stream **class** (each instance stores the counter value at the moment of construction, then increases it) - best IMO
  - pros: simple
  - cons: thread safety? I'm pretty sure class attributes access is atomic, though
- Each Node instance increases a counter whenever `.stream()` is called, and passes the number to Stream constructor.
  - pros: idk lol
  - cons: Node is not immutable any more, API changes

After one of these values is available, `outgoing_edge_map` can be an `OrderedDict` that is sorted according to the values.

Another way would be to sort the edges using their upstream label.

**Comment by Depaulicious** [comment]
Actually I think upstream label sorting makes more sense: I may want to do this:

```python
ref2 = s2ref[1]
scaled = s2ref[0]
```
and it should work.

**Comment by Depaulicious** [comment]
Okay, I implemented it with label sorting in #66. I used `OrderedDict` to force a _FIFO_ iteration. The dict is generated by the topological sorting algorithm, then re-sorted at the end (I don't see a way to generate it sorted).

**Comment by kkroening** [comment]
We need a test for this. I'll see if I can write one.

**Comment by kkroening** [comment]
I think there's a simpler approach:
```diff
def _allocate_filter_stream_names(filter_nodes, outgoing_edge_maps, stream_name_map):
    stream_count = 0
    for upstream_node in filter_nodes:
        outgoing_edge_map = outgoing_edge_maps[upstream_node]
-        for upstream_label, downstreams in list(outgoing_edge_map.items()):
+        for upstream_label, downstreams in sorted(outgoing_edge_map.items()):
            if len(downstreams) > 1:
                # TODO: automatically insert `splits` ahead of time via graph transformation.
                raise ValueError('Encountered {} with multiple outgoing edges with same upstream label {!r}; a '
                    '`split` filter is probably required'.format(upstream_node, upstream_label))
            stream_name_map[upstream_node, upstream_label] = _get_stream_name('s{}'.format(stream_count))
            stream_count += 1
```
That one-line fix should do the trick I think, but we'll be able to verify once we have a test that demonstrates the issue.

The test should fail on master and pass on your branch.

**Comment by kkroening** [comment]
Hmm, I can't reproduce the issue. Here's my test:
```
def test_outgoing_edge_sorting():
    in1 = ffmpeg.input('in1.mp4')
    in2 = ffmpeg.input('in2.mp4')
    s2ref = ffmpeg.filter_multi_output([in1, in2], 'scale2ref')
    out1 = s2ref[0].hflip().output('out1.mp4')
    out2 = s2ref[1].output('out2.mp4')
    out = ffmpeg.merge_outputs(out1, out2)
    args = out.get_args()
    assert args == [
        '-i', 'in1.mp4',
        '-i', 'in2.mp4',
        '-filter_complex', '[0][1]scale2ref[s0][s1];[s0]hflip[s2]',
        '-map', '[s2]', 'out1.mp4',
        '-map', '[s1]', 'out2.mp4',
    ]
```
I tried switching all sorts of stuff around to try to get `scale2ref` to `hflip` the wrong thing, but master + your branch both produce the same result no matter what I try.

Could you make a test case that fails on master but passes on your branch to demonstrate the issue?

**Comment by Depaulicious** [comment]
Yeah it's really hard to reproduce. We would need to find out when Python's dict implementation returns the elements out of order. I'll see how it's implemented later this week and try to make a test case.

**Comment by komar007** [comment]
Here's a unit test which reproduces the issue:

```
def test__multi_output_edge_label_order():
    scale2ref = ffmpeg.filter_multi_output([ffmpeg.input('x'), ffmpeg.input('y')], 'scale2ref')
    out = (
        ffmpeg.merge_outputs(
            scale2ref[1].filter('scale').output('a'),
            scale2ref[10000].filter('hflip').output('b')
        )
    )

    args = out.get_args()
    assert args == [
        '-i',
        'x',
        '-i',
        'y',
        '-filter_complex',
        '[0][1]scale2ref[s0][s1];[s0]scale[s2];[s1]hflip[s3]',
        '-map',
        '[s2]',
        'a',
        '-map',
        '[s3]',
        'b'
    ]

```

Python has optimizations for small numbers and short strings, so it's likely that dicts with keys like 1, 2 or 3 will enumerate in ascending order. Use 1 and 10000 and the order is no longer maintained.

@kkroening the simpler solution you proposed will not work (at least now, in 2019, because I haven't traced the changes in the code since your comment), you'd also have to fix `get_incoming_edges` in the same way.

@Depau 's proposition fixes the issue by returning the edges sorted correctly from `topo_sort`. Assuming it's okay to put that responsibility on `topo_sort`, I have a simpler solution, let's return lists of tuples instead - I have to verify this, but it looks like the dicts are currently only iterated and never accessed randomly - so there should be no performance penalty.
I can publish my changes tomorrow.

If you don't think it should be `topo_sort`'s responsibility to sort multiple outgoing edges by label, then I have verified that iterating sorted `items()` both in `_allocate_filter_stream_names` and `get_incoming_edges` fixes the issue.

**Comment by 153957** [comment]
Is this not also fixed by using Python 3.6+ ?
Of course Python 2.7 is still officially supported for half a year, and python 3.5 for a bit over a year..

**Comment by komar007** [comment]
@153957 Python 3.6 maintains insertion order of dict elements, not order based on comparison of keys.
Still, it's an implementation detail of python 3.6 and insertion order is only guaranteed by spec from python 3.7 onwards.

I think we agree that in case of nodes returned from `.stream(x)` we should not take order of any calls into consideration as that could case codes like:
```
out1, out2 = filter.stream(0), filter.stream(1)
```
to behave differently than:
```
out2, out1 = filter.stream(1), filter.stream(0)

It seems to me that label values are currently the only caller-provided source of information about the intentional order of outputs from multiple-output filters.
Currently, the labels are never compared explicitly, making the order of outputs of multiple-output filters completely implementation-dependent, as shown by the example test I provided.

I hope this clears things up a bit.
